### PR TITLE
Introduce SamplingMechanism param to setSamplingPriority

### PR DESCRIPTION
**Pull request by ygree**
# What Does This Do
Modifies DDSpanContext::setSamplingPriority to have a new parameter to track the mechanism which is setting/changing the sampling priority. 

# Motivation
Trace Tags Propagation in Distributed Traces RFC introduces sampling mechanism.

# Additional Notes
[Implementation plan](https://docs.google.com/document/d/1eBOz7BwCEDc6yuM58yMsNN3Ov0jbjRnZcXgYs2zjuTw)
See “Sampling Mechanisms” section of [the RFC](https://docs.google.com/document/d/1zeO6LGnvxk5XweObHAwJbK3SfK23z7jQzp7ozWJTa2A).

### CODE REVIEW

File: dd-trace-core/src/main/java/datadog/trace/common/writer/ListWriter.java [hunk 1]
```diff
@@ -66,7 +66,7 @@ public boolean waitForTracesMax(final int number, int seconds)
   }
 
   public void waitForTraces(final int number) throws InterruptedException, TimeoutException {
-    if (!waitForTracesMax(number, 20)) {
+    if (!waitForTracesMax(number, 200)) {
```
Comment (bantonsson) [inline dd-trace-core/src/main/java/datadog/trace/common/writer/ListWriter.java:69]: Waiting for `200` seconds seems a bit excessive.

File: dd-trace-core/src/main/java/datadog/trace/core/DDSpan.java [hunk 2]
```diff
@@ -528,20 +529,28 @@ public void finishWork() {
     return context.getRequestContext();
   }
 
+  @Deprecated
+  @Override
+  public final DDSpan setSamplingPriority(final int newPriority) {
+    // this method exist only to satisfy MutableSpan. It will be removed in 1.0
```
Comment (bantonsson) [inline dd-trace-core/src/main/java/datadog/trace/core/DDSpan.java:535]: 👍

File: dd-trace-core/src/main/java/datadog/trace/core/DDSpanContext.java [hunk 3]
```diff
@@ -266,23 +266,32 @@ public void forceKeep() {
     }
     // if the user really wants to keep this trace chunk, we will let them,
     // even if the old sampling priority has already propagated
-    SAMPLING_PRIORITY_UPDATER.set(this, USER_KEEP);
+    SAMPLING_PRIORITY_UPDATER.set(this, PrioritySampling.USER_KEEP);
```
Comment (bantonsson) [inline dd-trace-core/src/main/java/datadog/trace/core/DDSpanContext.java:269]: I think this needs to be rewritten in some way to account for the sampling mechanism as well.

File: dd-trace-core/src/main/java/datadog/trace/core/DDSpanContext.java [hunk 4]
```diff
@@ -266,23 +266,32 @@ public void forceKeep() {
...
       return false;
     }
 
+    if (!SamplingMechanism.validateWithSamplingPriority(newMechanism, newPriority)) {
+      log.debug(
+          "{}: Refusing to set samplingMechanism to {}. Provided samplingPriority {} is not allowed.",
+          this,
+          newMechanism,
+          newPriority);
+      return false;
```
Comment (bantonsson) [inline dd-trace-core/src/main/java/datadog/trace/core/DDSpanContext.java:285]: Should we have a configurable fallback mode where we set the priority anyway, if this change breaks broken customer code?

**Review by bantonsson — COMMENTED** [review]
Looks good overall. Only some minor comments/questions.

File: dd-trace-core/src/main/java/datadog/trace/common/writer/ListWriter.java (hunk 1 above)
Comment (ygree) [reply dd-trace-core/src/main/java/datadog/trace/common/writer/ListWriter.java:69]: Good catch. I had aerospike and mule tests failing locally and was trying to increase this time out.

File: dd-trace-core/src/main/java/datadog/trace/core/DDSpanContext.java (hunk 4 above)
Comment (ygree) [reply dd-trace-core/src/main/java/datadog/trace/core/DDSpanContext.java:285]: Good idea! I'll add a config parameter for that. Thanks!

File: dd-trace-core/src/main/java/datadog/trace/core/DDSpan.java (hunk 2 above)
Comment (ygree) [reply dd-trace-core/src/main/java/datadog/trace/core/DDSpan.java:535]: Done

File: dd-trace-core/src/main/java/datadog/trace/core/DDSpanContext.java (hunk 3 above)
Comment (ygree) [reply dd-trace-core/src/main/java/datadog/trace/core/DDSpanContext.java]: Done

File: dd-trace-core/src/main/java/datadog/trace/core/DDSpanContext.java (hunk 4 above)
Comment (ygree) [reply dd-trace-core/src/main/java/datadog/trace/core/DDSpanContext.java]: Done

**Review by bantonsson — APPROVED** [review]

File: dd-trace-api/src/main/java/datadog/trace/api/config/TracerConfig.java [hunk 5]
```diff
@@ -58,6 +58,8 @@
   public static final String PROPAGATION_STYLE_INJECT = "propagation.style.inject";
 
   public static final String ENABLE_TRACE_AGENT_V05 = "trace.agent.v0.5.enabled";
+  public static final String SAMPLING_MECHANISM_VALIDATION_DISABLED =
+      "trace.sampling.mechanism.validation";
```
Comment (bantonsson) [inline dd-trace-api/src/main/java/datadog/trace/api/config/TracerConfig.java:62]: Maybe this should have the full wording, i.e. `"trace.sampling.mechanism.validation.disabled"`

Comment (ygree) [reply dd-trace-api/src/main/java/datadog/trace/api/config/TracerConfig.java]: Good idea! I've changed according to your suggestion.

File: internal-api/src/main/java/datadog/trace/api/sampling/SamplingMechanism.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  /** Auto; reserved for future use */
+  public static final byte REMOTE_AUTO_RATE = 2;
+  /** Sampling rule or sampling rate based on tracer config */
+  public static final byte RULE = 3;
+  /** User directly sets sampling priority via code using span.SetTag(ManualKeep) or similar API */
+  public static final byte MANUAL = 4;
+  /** AppSec */
+  public static final byte APPSEC = 5;
+  /** User-defined target; reserved for future use */
+  public static final byte REMOTE_USER_RATE = 6;
```
Comment (bantonsson) [inline internal-api/src/main/java/datadog/trace/api/sampling/SamplingMechanism.java:21]: I just noticed while reading through some RFCs that there is a `REMOTE_EMERGENCY_RATE = 7` missing from this, or is it left out on purpose?

Comment (ygree) [reply internal-api/src/main/java/datadog/trace/api/sampling/SamplingMechanism.java:21]: Thanks! The spec doesn't specify allowed priorities for `REMOTE_EMERGENCY_RATE` or how this mechanism can be used. I've added a question to the spec: https://docs.google.com/document/d/1zeO6LGnvxk5XweObHAwJbK3SfK23z7jQzp7ozWJTa2A/edit?disco=AAAASkzXaI4

**Pull request merged**